src/modules/types.rs

```diff
@@ -14,16 +14,35 @@ pub enum Type {
     Generic
 }
 
+impl Type {
+    pub fn is_subset_of(&self, of: &Type) -> bool {
+        match (self, of) {
+            (Type::Array(left), Type::Array(right)) => {
+                **left != Type::Generic && **right == Type::Generic
+            }
+            (left, Type::Failable(right)) => !matches!(left, Type::Failable(_)) && (*left == **right || left.is_subset_of(right)),
+            _ => false
```

Comment:
Should `Text` is subset of `Text` be equal to `false`? Mathematically this should be true as we are comparing two singletons:

<img width="296" alt="image" src="https://github.com/user-attachments/assets/aa16f181-3aed-4faf-bcb4-5792eae9d75c">

Reply:
This is supposed to be the proper/strict subset

![Screenshot_20240915-145854.png](https://github.com/user-attachments/assets/19269aa5-f247-46f2-a5e7-8bcd9a4bf3f1)

Reply:
I though that this function is supposed to check if type B can be narrowed down to type A. In this case it is not a strict subset:
```py
Text | Num => Text    True
Bool | Num => Text    False
Text       => Text    True
Bool       => Text    False
```

Reply:
Even if you put it this way, "narrowing down" implies a change. If the types are the same from the beginning, there is no change. This is bike shedding, though. 

I prefer to keep this function strictly about checking for the proper subset, instead of making it a "can this type pass" function. This way it is more universal — we can differentiate whether the types are the same, or sub/superset of each other (if we ever need to).

This also has an additional benefit of not executing unless necessary — type checking in args/return will short-circuit on a simple equality check, unless the types aren't the same. 

If you want to deduplicate the type checking logic from args/return, we can put it in a separate function, that will most likely get inlined (and we can even hint the compiler to do so). 

Although I don't think this logic is complicated enough to warrant a separate function. 

FYI https://nnethercote.github.io/perf-book/inlining.html#simple-cases